File: app/src/main/java/tw/firemaples/onscreenocr/views/OcrResultWindow.java
package tw.firemaples.onscreenocr.views;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.util.DisplayMetrics;
import android.view.Display;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.RelativeLayout;
import android.widget.TextView;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Locale;

import tw.firemaples.onscreenocr.R;
import tw.firemaples.onscreenocr.floatingviews.screencrop.TTSPlayerView;
import tw.firemaples.onscreenocr.ocr.OcrNTranslateState;
import tw.firemaples.onscreenocr.ocr.OcrResult;
import tw.firemaples.onscreenocr.utils.FabricUtil;
import tw.firemaples.onscreenocr.utils.OcrNTranslateUtils;
import tw.firemaples.onscreenocr.utils.SharePreferenceUtil;
import tw.firemaples.onscreenocr.utils.Tool;
import tw.firemaples.onscreenocr.utils.ViewPreparedWaiter;

/**
 * Created by firemaples on 29/11/2016.
 */

public class OcrResultWindow {
    private static final Logger logger = LoggerFactory.getLogger(OcrResultWindow.class);

    private final static int MARGIN = 10;

    private final Context context;
    private ViewGroup parent;
    private View anchorView;

    private View rootView;
    private View view_translatedTextWrapper;
    private View pb_origin, pb_translated;
    private View bt_openInBrowser_ocrText, bt_copy_ocrText, bt_edit_ocrText, bt_tts_ocrText, bt_openGoogleTranslate_ocrText;
    private View bt_openInBrowser_translatedText, bt_copy_translatedText, bt_tts_translatedText, bt_openGoogleTranslate_translatedText;
    private TextView tv_originText, tv_translatedText;
    private RelativeLayout.LayoutParams layoutParams;
    private DisplayMetrics metrics;

    private OnOcrResultWindowCallback callback;

    private OcrNTranslateState state;
    private OcrResult ocrResult;

    public OcrResultWindow(Context context, ViewGroup parent, OnOcrResultWindowCallback callback) {
        this.context = context;
        this.parent = parent;
        this.callback = callback;

        rootView = View.inflate(context, R.layout.view_ocr_result_window, null);

        view_translatedTextWrapper = rootView.findViewById(R.id.view_translatedTextWrapper);

        pb_origin = rootView.findViewById(R.id.pb_origin);
        pb_translated = rootView.findViewById(R.id.pb_translated);

        tv_originText = (TextView) rootView.findViewById(R.id.tv_originText);
        tv_translatedText = (TextView) rootView.findViewById(R.id.tv_translatedText);

        bt_tts_ocrText = rootView.findViewById(R.id.bt_tts_ocrText);
        bt_edit_ocrText = rootView.findViewById(R.id.bt_edit_ocrText);
        bt_copy_ocrText = rootView.findViewById(R.id.bt_copy_ocrText);
        bt_openInBrowser_ocrText = rootView.findViewById(R.id.bt_openInBrowser_ocrText);
        bt_openGoogleTranslate_ocrText = rootView.findViewById(R.id.bt_openGoogleTranslate_ocrText);

        bt_tts_translatedText = rootView.findViewById(R.id.bt_tts_translatedText);
        bt_copy_translatedText = rootView.findViewById(R.id.bt_copy_translatedText);
        bt_openInBrowser_translatedText = rootView.findViewById(R.id.bt_openInBrowser_translatedText);
        bt_openGoogleTranslate_translatedText = rootView.findViewById(R.id.bt_openGoogleTranslate_translatedText);

        bt_tts_ocrText.setOnClickListener(onClickListener);
        bt_edit_ocrText.setOnClickListener(onClickListener);
        bt_copy_ocrText.setOnClickListener(onClickListener);
        bt_openInBrowser_ocrText.setOnClickListener(onClickListener);
        bt_openGoogleTranslate_ocrText.setOnClickListener(onClickListener);

        bt_copy_translatedText.setOnClickListener(onClickListener);
        bt_openInBrowser_translatedText.setOnClickListener(onClickListener);
        bt_tts_translatedText.setOnClickListener(onClickListener);
        bt_openGoogleTranslate_translatedText.setOnClickListener(onClickListener);

        view_translatedTextWrapper.setVisibility(SharePreferenceUtil.getInstance().isEnableTranslation() ? View.VISIBLE : View.GONE);

        layoutParams = new RelativeLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        layoutParams.setMargins(MARGIN, MARGIN, MARGIN, MARGIN);

        WindowManager wm = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        Display display = wm.getDefaultDisplay();
        metrics = new DisplayMetrics();
        display.getMetrics(metrics);
    }

    public void setOcrResult(OcrNTranslateState state, OcrResult ocrResult) {
        this.state = state;
        this.ocrResult = ocrResult;

//        switch (state) {
//            case OCR_INIT:
//            case OCR_RUNNING:
//                pb_origin.setVisibility(View.VISIBLE);
//                pb_translated.setVisibility(View.GONE);
//                break;
//            case OCR_FINISHED:
//                pb_origin.setVisibility(View.GONE);
//                tv_originText.setText(ocrResult.getText());
//                break;
//            case TRANSLATING:
//                pb_translated.setVisibility(View.VISIBLE);
//                tv_originText.setText(ocrResult.getText());
//                break;
//            case TRANSLATED:
//                pb_translated.setVisibility(View.GONE);
//                tv_originText.setText(ocrResult.getText());
//                tv_translatedText.setText(ocrResult.getTranslatedText());
//                break;
//        }

        boolean ocrFinished = state.getStep() >= OcrNTranslateState.OCR_FINISHED.getStep();
        boolean translated = state.getStep() >= OcrNTranslateState.TRANSLATED.getStep();

        pb_origin.setVisibility(!ocrFinished ? View.VISIBLE : View.GONE);
        pb_translated.setVisibility(ocrFinished && !translated ? View.VISIBLE : View.GONE);
        if (ocrFinished) {
            tv_originText.setText(ocrResult.getText());
        }
        if (translated) {
            tv_translatedText.setText(ocrResult.getTranslatedText());
        }

        bt_tts_ocrText.setEnabled(ocrFinished);
        bt_edit_ocrText.setEnabled(ocrFinished);
        bt_copy_ocrText.setEnabled(ocrFinished);
        bt_openInBrowser_ocrText.setEnabled(ocrFinished);
        bt_openGoogleTranslate_ocrText.setEnabled(ocrFinished);

        bt_tts_translatedText.setEnabled(translated);
        bt_copy_translatedText.setEnabled(translated);
        bt_openInBrowser_translatedText.setEnabled(translated);
        bt_openGoogleTranslate_translatedText.setEnabled(translated);
    }

    public void show(View anchorView) {
        dismiss();
        this.anchorView = anchorView;

        new ViewPreparedWaiter().waitView(rootView, onViewPrepared);
        parent.addView(rootView, layoutParams);
    }

    public void dismiss() {
        if (rootView.getParent() != null) {
            ((ViewGroup) rootView.getParent()).removeView(rootView);
        }
    }

    private void adjustViewPosition() {
        int width = rootView.getWidth();
        int height = rootView.getHeight();

        int parentHeight = parent.getHeight();

        int needHeight = height + MARGIN * 2;

        if (anchorView.getTop() > needHeight) {
            //Gravity = TOP
            layoutParams.topMargin = anchorView.getTop() - height;
        } else if (parentHeight - anchorView.getBottom() > needHeight) {
            //Gravity = BOTTOM
            layoutParams.topMargin = anchorView.getTop() + anchorView.getHeight();
        } else {
            layoutParams.topMargin = MARGIN;
            layoutParams.addRule(RelativeLayout.CENTER_VERTICAL);
        }

        if (anchorView.getLeft() + width + MARGIN > metrics.widthPixels) {
            // Match screen right
            layoutParams.leftMargin = metrics.widthPixels - (width + MARGIN);
        } else {
            // Match anchorView left
            layoutParams.leftMargin = anchorView.getLeft();
        }

        parent.updateViewLayout(rootView, layoutParams);
    }

    private ViewPreparedWaiter.OnViewPrepared onViewPrepared = new ViewPreparedWaiter.OnViewPrepared() {
        @Override
        public void onViewPrepared(View viewToWait) {
            logger.info("Width:" + viewToWait.getWidth() + " Height:" + viewToWait.getHeight());
            adjustViewPosition();
        }
    };

    private View.OnClickListener onClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            int id = v.getId();
            if (id == R.id.bt_openInBrowser_ocrText) {
                FabricUtil.logBtnOpenInWebViewClicked("OCR text");
                if (ocrResult != null && ocrResult.getText() != null) {
                    callback.onOpenBrowserBtnClick(ocrResult.getText(), false);
                }
            } else if (id == R.id.bt_openInBrowser_translatedText) {
                FabricUtil.logBtnOpenInWebViewClicked("Translated text");
                if (ocrResult != null && ocrResult.getTranslatedText() != null) {
                    callback.onOpenBrowserBtnClick(ocrResult.getTranslatedText(), true);
                }
            } else if (id == R.id.bt_copy_ocrText) {
                copyToClipboard("OCR text", ocrResult.getText());
            } else if (id == R.id.bt_copy_translatedText) {
                copyToClipboard("Translated text", ocrResult.getTranslatedText());
            } else if (id == R.id.bt_edit_ocrText) {
                if (ocrResult != null) {
                    callback.onEditOriTextClicked(ocrResult);
                }
            } else if (id == R.id.bt_tts_ocrText || id == R.id.bt_tts_translatedText) {
                String lang;
                String ttsContent;
                if (id == R.id.bt_tts_ocrText) {
                    FabricUtil.logBtnPlayTTSClicked("OCR text");
                    lang = OcrNTranslateUtils.getInstance().getTranslateFromLang();
                    ttsContent = ocrResult.getText();
                } else {
                    FabricUtil.logBtnPlayTTSClicked("Translated text");
                    lang = OcrNTranslateUtils.getInstance().getTranslateToLang();
                    ttsContent = ocrResult.getTranslatedText();
                }

                TTSPlayerView ttsPlayerView = new TTSPlayerView(context);
                ttsPlayerView.setTTSContent(lang, ttsContent);
                ttsPlayerView.attachToWindow();
            } else if (id == R.id.bt_openGoogleTranslate_ocrText) {
                if (ocrResult != null && ocrResult.getText() != null) {
                    callback.openGoogleTranslate(ocrResult.getText(), false);
                }
            } else if (id == R.id.bt_openGoogleTranslate_translatedText) {
                if (ocrResult != null && ocrResult.getTranslatedText() != null) {
                    callback.openGoogleTranslate(ocrResult.getTranslatedText(), true);
                }
            }
        }
    };

    private void copyToClipboard(String label, String text) {
        FabricUtil.logBtnCopyToClipboardClicked(label);
        ClipboardManager clipboard = (ClipboardManager)
                context.getSystemService(Context.CLIPBOARD_SERVICE);
        ClipData clipData = ClipData.newPlainText(label, text);
        clipboard.setPrimaryClip(clipData);
        Tool.getInstance().showMsg(String.format(Locale.getDefault(), context.getString(R.string.msg_textHasBeenCopied), text));
    }

    public interface OnOcrResultWindowCallback {
        void onOpenBrowserBtnClick(String text, boolean translated);

        void onEditOriTextClicked(OcrResult ocrResult);

        void openGoogleTranslate(String text, boolean translated);
    }
}

File: app/src/main/java/tw/firemaples/onscreenocr/views/OCRResultWindow.kt
package tw.firemaples.onscreenocr.views

import android.content.Context
import android.graphics.Rect
import android.util.DisplayMetrics
import android.view.View
import android.view.View.OnClickListener
import android.view.ViewGroup
import android.view.WindowManager
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.RelativeLayout
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import org.slf4j.LoggerFactory
import tw.firemaples.onscreenocr.R
import tw.firemaples.onscreenocr.StateManager
import tw.firemaples.onscreenocr.event.EventUtil
import tw.firemaples.onscreenocr.floatings.screencrop.TTSPlayerView
import tw.firemaples.onscreenocr.floatings.screencrop.TextEditDialogView
import tw.firemaples.onscreenocr.ocr.tesseract.OCRLangUtil
import tw.firemaples.onscreenocr.state.TranslatedState
import tw.firemaples.onscreenocr.state.TranslatingState
import tw.firemaples.onscreenocr.state.event.StateChangedEvent
import tw.firemaples.onscreenocr.translate.GoogleTranslateUtil
import tw.firemaples.onscreenocr.translate.TranslationUtil
import tw.firemaples.onscreenocr.utils.*

class OCRResultWindow(context: Context) : FrameLayout(context) {
    private val logger = LoggerFactory.getLogger(OCRResultWindow::class.java)

    private val MARGIN_PX = 4f
    private val layoutMargin by lazy { UIUtil.dpToPx(context, MARGIN_PX) }

    private val view: View by lazy { View.inflate(context, R.layout.view_ocr_result_window, null) }

    private val pbOri by lazy { view.getView(R.id.pb_origin) }
    private val tvOriText by lazy { view.getTextView(R.id.tv_originText) }
    private val btCopyOT by lazy { view.getView(R.id.bt_copy_ocrText) }
    private val btEditOT by lazy { view.getView(R.id.bt_edit_ocrText) }
    private val btTtsOT by lazy { view.getView(R.id.bt_tts_ocrText) }
    private val btOpenGtOT by lazy { view.getView(R.id.bt_openGoogleTranslate_ocrText) }

    private val wrapperTranslatedArea by lazy { view.getView(R.id.view_translatedTextWrapper) }
    private val pbTranslated by lazy { view.getView(R.id.pb_translated) }
    private val tvTranslated by lazy { view.getTextView(R.id.tv_translatedText) }
    private val btCopyTT by lazy { view.getView(R.id.bt_copy_translatedText) }
    private val btTtsTT by lazy { view.getView(R.id.bt_tts_translatedText) }
    private val btOpenGtTT by lazy { view.getView(R.id.bt_openGoogleTranslate_translatedText) }

    private val tvTranslationService by lazy { view.getTextView(R.id.tv_translationService) }
    private val ivTranslationService by lazy { view.findViewById<ImageView>(R.id.iv_translationService) }

    private val layoutParams by lazy {
        RelativeLayout.LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT,
                ViewGroup.LayoutParams.WRAP_CONTENT).apply {
            setMargins(layoutMargin, layoutMargin, layoutMargin, layoutMargin)
        }
    }

    private val displayMetrics by lazy {
        val metrics = DisplayMetrics()
        (context.getSystemService(Context.WINDOW_SERVICE) as WindowManager)
                .defaultDisplay.getMetrics(metrics)
        metrics
    }

    private val onClickListener: OnClickListener = OnClickListener { v ->
        when (v) {
            btTtsOT -> {
                showTTSPlayer(OCRLangUtil.selectLangDisplayCode,
                        tvOriText.text.toString())
            }
            btTtsTT -> {
                showTTSPlayer(TranslationUtil.currentTranslationLangCode,
                        tvTranslated.text.toString())
            }
            btCopyOT -> {
                Utils.copyToClipboard(Utils.LABEL_OCR_RESULT, tvOriText.text.toString())
            }
            btCopyTT -> {
                Utils.copyToClipboard(Utils.LABEL_TRANSLATED_TEXT, tvTranslated.text.toString())
            }
            btEditOT -> {
                changeOCRText()
            }
            btOpenGtOT -> {
                openGoogleTranslateApp(TranslationUtil.currentTranslationLangCode,
                        tvOriText.text.toString())
            }
            btOpenGtTT -> {
                openGoogleTranslateApp(OCRLangUtil.selectLangDisplayCode,
                        tvTranslated.text.toString())
            }
        }
    }

    init {
        isClickable = true
        isFocusable = true

        addView(view)
        btCopyOT.setOnClickListener(onClickListener)
        btEditOT.setOnClickListener(onClickListener)
        btTtsOT.setOnClickListener(onClickListener)
        btOpenGtOT.setOnClickListener(onClickListener)
        btCopyTT.setOnClickListener(onClickListener)
        btTtsTT.setOnClickListener(onClickListener)
        btOpenGtTT.setOnClickListener(onClickListener)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        EventUtil.register(this)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        EventUtil.unregister(this)
    }

    @Suppress("unused")
    @Subscribe(threadMode = ThreadMode.MAIN, sticky = true)
    fun onStateChanged(event: StateChangedEvent) {
        wrapperTranslatedArea.setVisible(TranslationUtil.isEnableTranslation)

        val ocrFinished = event.state in arrayOf(TranslatingState, TranslatedState)
        val translated = event.state == TranslatedState

        pbOri.visibility = if (ocrFinished) View.GONE else View.VISIBLE
        pbTranslated.visibility = if (translated) View.GONE else View.VISIBLE

        var oriText: String? = null
        var translatedText: String? = null
        if (StateManager.ocrResultList.isNotEmpty()) {
            val ocrResult = StateManager.ocrResultList[0]
            oriText = if (ocrFinished) ocrResult.text else null
            translatedText = if (translated) ocrResult.translatedText else null
        }
        tvOriText.text = oriText
        tvTranslated.text = translatedText

        btCopyOT.isEnabled = ocrFinished
        btEditOT.isEnabled = ocrFinished
        btTtsOT.isEnabled = ocrFinished
        btOpenGtOT.isEnabled = ocrFinished

        btCopyTT.isEnabled = translated
        btTtsTT.isEnabled = translated
        btOpenGtTT.isEnabled = translated

        val service = TranslationUtil.currentService
        if (service.resultDrawableResId != -1) {
            ivTranslationService.setImageResource(service.resultDrawableResId)
            ivTranslationService.setVisible(true)
            tvTranslationService.setVisible(false)
        } else {
            val translatedBy = context.getString(
                    R.string.translatedBy, service.fullName)
            tvTranslationService.text = translatedBy
            tvTranslationService.setVisible(TranslationUtil.isEnableTranslation)
            ivTranslationService.setVisible(false)
        }
    }

    private var anchorView: View? = null
    fun anchor(anchorView: View) {
        this.anchorView = anchorView

        view.onViewPrepared {
            logger.info("onViewPrepared, width: ${it.width}, height: ${it.height}")
            adjustViewPosition()
        }
    }

    private fun adjustViewPosition1() {
        val anchorView = this.anchorView ?: return
        val parent = this.parent as ViewGroup

        val width = this.width
        val height = this.height

        val parentHeight = parent.height

        val needHeight = height + layoutMargin * 2

        if (anchorView.top > needHeight) {
            //Gravity = TOP
            layoutParams.topMargin = anchorView.top - height
        } else if (parentHeight - anchorView.bottom > needHeight) {
            //Gravity = BOTTOM
            layoutParams.topMargin = anchorView.top + anchorView.height
        } else {
            layoutParams.topMargin = layoutMargin
            layoutParams.addRule(RelativeLayout.CENTER_VERTICAL)
        }

        if (anchorView.left + width + layoutMargin > displayMetrics.widthPixels) {
            // Match screen right
            layoutParams.leftMargin = displayMetrics.widthPixels - (width - layoutMargin)
        } else {
            // Match anchorView left
            layoutParams.leftMargin = anchorView.left
        }

        parent.updateViewLayout(this, layoutParams)
    }

    private fun adjustViewPosition() {
        val anchorView = this.anchorView ?: return
        val parent = this.parent as ViewGroup

        val width = this.width
        val height = this.height

        val parentHeight = parent.height

        val rect = Rect()
        anchorView.getGlobalVisibleRect(rect)

        val margins = UIUtil.countViewPosition(rect, width, height, layoutMargin, parentHeight)

        layoutParams.leftMargin = margins[0]
        layoutParams.topMargin = margins[1]

        if (layoutParams.topMargin == -1) {
            layoutParams.topMargin = layoutMargin
            layoutParams.addRule(RelativeLayout.CENTER_VERTICAL)
        }

        parent.updateViewLayout(this, layoutParams)
    }

    private fun showTTSPlayer(lang: String, text: String) {
        TTSPlayerView(context).apply {
            setTTSContent(lang, text)
            attachToWindow()
        }
    }

    private fun changeOCRText() {
        TextEditDialogView(context).apply {
            setTitle(context.getString(R.string.title_editOCRText))
            setContentText(tvOriText.text.toString())
            setCallback(object : TextEditDialogView.OnTextEditDialogViewCallback() {
                override fun onConfirmClick(textEditDialogView: TextEditDialogView?, text: String?) {
                    super.onConfirmClick(textEditDialogView, text)
                    if (text != null) {
                        StateManager.changeOCRText(text)
                    }
                }
            })

            attachToWindow()
        }
    }

    private fun openGoogleTranslateApp(langTo: String, text: String) {
        GoogleTranslateUtil.start(context, langTo, text)
        StateManager.onBackButtonPressed()
    }
}